clang-tools-extra/clangd/XRefs.cpp

```diff
@@ -1354,6 +1354,8 @@ maybeFindIncludeReferences(ParsedAST &AST, Position Pos,
 
         ReferencesResult::Reference Result;
         const auto *Token = AST.getTokens().spelledTokenAt(Loc);
+        if (!Token)
```

Comment:
it's a little bit hard to see what issue we're fixing/preventing here, can you add some context?

Looking at the test case, i guess we're getting a reference for the UDLs pointing at the middle of a token?

If that's the case, i think we might as well fix the feature instead of working around. Moreover I don't think that broken behavior is specific to xrefs on includes, it'll show up on other places that use `spelledTokenAt` too.

In this code path particularly, we actually don't want a `Token`, we can just use `Loc` and `Lexer::getLocForEndOfToken(Loc)` for the `range` calculation below.

I think a more complete fix would be to actually change `spelledTokenAt` to return the `Token` as long as `Loc` is contained (or have a new method that returns the token containing the location). That one is more involved though, so if you choose not to do that, can you at least file a bug around clangd features not working properly with UDLs pointing into this PR?

Reply:
I have filled https://github.com/llvm/llvm-project/issues/94599.

```cpp
bool operator"" _b(unsigned long long value);
auto x = 1_b;
```
As you said, the problem is that we get a reference to the middle of a spelled token. Ref at`1^_b`, spelled token: `[[1^b]]`.

All the users of `spelledTokenAt` seems to be doing the null check or an assertion. So I think a null check here seems warranted (if not an assertion).

As far as supporting UDL as `spelledTokens`, in general, I don't have a great idea. `clang::Token` from `Lexer::LexFromRawLexer` seems to not break the token into two [1][_b].

Another simpler solution could be to just change `spelledTokens` and instead of returning `nullptr` in 
````
 if (!Tok || Tok->location() != Loc)
    return nullptr;
```
we could return if `Tok` if `Tok->location() <= Loc && Loc < Tok->endLocation()`
WDYT ?

Reply:
> All the users of spelledTokenAt seems to be doing the null check or an assertion. So I think a null check here seems warranted (if not an assertion).

this place also deserves an assertion, we already go the extra mile to map the location into main-file, hence we can also assert (but that won't help with your crash :/)

> As far as supporting UDL as spelledTokens, in general, I don't have a great idea. clang::Token from Lexer::LexFromRawLexer seems to not break the token into two [1][_b].

Unfortunately it's a single token for raw-lexer. But we don't need a token here, it's enough to get the end location for the token. Hence I was suggesting to use `Lexer::getLocForEndOfToken` and just populate the ranges, without really finding a `Token` associated with the location. (or are you saying even `getLocForEndOfToken` is doing the wrong thing here?)

> we could return if Tok if Tok->location() <= Loc && Loc < Tok->endLocation()

Right, that's what I meant with `I think a more complete fix would be to actually change spelledTokenAt to return the Token as long as Loc is contained`. If you feel like it, i'd definitely appreciate such a change. All call-sites are in clangd, so we just need to decide if all of them are actually OK with this behavior and don't really rely on `spelledTokenAt` failing when it's called with a random offset into the token.

Reply:
Modified `spelledTokenAt`.

Reply:
thanks! can you also rename `spelledTokenAt` to `spelledTokenCotaining` and update the comment and change the null-check here to an assertion?

also, what's the conclusion of investigation around `don't really rely on spelledTokenAt failing when it's called with a random offset into the token`?

Reply:
Done.
> also, what's the conclusion of investigation around don't really rely on spelledTokenAt failing when it's called with a random offset into the token?

I relied on the tests here. Also, the usages of `spelledTokenCotaining` does not seem to relying on getting a `nullptr` if the location is inside a token, as compared to strictly at the beginning of tok.